### PR TITLE
Let `openssl crl` support the use of engines.

### LINKED ISSUE #11751: the crl app does not support -engine command line parameter
Here is some code that can do it.  
If its satisfactory and a desired thing to do, I can open a PR: 

```
diff --git a/openssl-1.1.1g/apps/crl.c b/openssl-1.1.1g/apps/crl.c
index 031fada1..2b5e1480 100644
--- a/openssl-1.1.1g/apps/crl.c
+++ b/openssl-1.1.1g/apps/crl.c
@@ -24,7 +24,7 @@ typedef enum OPTION_choice {
     OPT_ISSUER, OPT_LASTUPDATE, OPT_NEXTUPDATE, OPT_FINGERPRINT,
     OPT_CRLNUMBER, OPT_BADSIG, OPT_GENDELTA, OPT_CAPATH, OPT_CAFILE,
     OPT_NOCAPATH, OPT_NOCAFILE, OPT_VERIFY, OPT_TEXT, OPT_HASH, OPT_HASH_OLD,
-    OPT_NOOUT, OPT_NAMEOPT, OPT_MD
+    OPT_NOOUT, OPT_NAMEOPT, OPT_MD, OPT_ENGINE
 } OPTION_CHOICE;

 const OPTIONS crl_options[] = {
@@ -56,6 +56,9 @@ const OPTIONS crl_options[] = {
     {"", OPT_MD, '-', "Any supported digest"},
 #ifndef OPENSSL_NO_MD5
     {"hash_old", OPT_HASH_OLD, '-', "Print old-style (MD5) hash value"},
+#endif
+#ifndef OPENSSL_NO_ENGINE
+    {"engine", OPT_ENGINE, 's', "Use engine, possibly a hardware device"},
 #endif
     {NULL}
 };
@@ -78,6 +81,7 @@ int crl_main(int argc, char **argv)
     int ret = 1, num = 0, badsig = 0, fingerprint = 0, crlnumber = 0;
     int text = 0, do_ver = 0, noCAfile = 0, noCApath = 0;
     int i;
+    ENGINE *e = NULL;
 #ifndef OPENSSL_NO_MD5
     int hash_old = 0;
 #endif
@@ -174,6 +178,9 @@ int crl_main(int argc, char **argv)
         case OPT_MD:
             if (!opt_md(opt_unknown(), &digest))
                 goto opthelp;
+        case OPT_ENGINE:
+            e = setup_engine(opt_arg(), 0);
+            break;
         }
     }
     argc = opt_num_rest();
@@ -338,5 +345,6 @@ int crl_main(int argc, char **argv)
     X509_CRL_free(x);
     X509_STORE_CTX_free(ctx);
     X509_STORE_free(store);
+    release_engine(e);
     return ret;
 }

```

### PR DESCRIPTION
**Pull request by anhu**
Fixes #11751

### CODE REVIEW

**Comment by anhu** [comment]
fixed the unhandled OPT_ENGINE in switch warning and all the other problems are unrelated to my changes.  Please review and if appropriate pull into master.

**Comment by beldmit** [comment]
At least the `check_docs` failure is relevant

**Comment by anhu** [comment]
> At least the `check_docs` failure is relevant

Indeed it is!! I'll get that fixed up.

File: apps/crl.c [hunk 1]
```diff
@@ -195,6 +201,11 @@ int crl_main(int argc, char **argv)
         case OPT_MD:
             digestname = opt_unknown();
             break;
+        case OPT_ENGINE:
+#ifndef OPENSSL_NO_ENGINE
+            e = setup_engine(opt_arg(), 0);
+#endif
```
Comment (beldmit) [inline apps/crl.c:207]: I see this call without #ifdef in other apps...

**Review by beldmit — COMMENTED** [review]
I don't see any usage of the engine in your code. Could you please explain how it gets used?

File: apps/crl.c (hunk 1 above)
Comment (anhu) [reply apps/crl.c:207]: `setup_engine()` will get your engine registered.  Then when you go to use algorithms your pmeth's and ameth's in your engine can  be selected.  You might think this looks strange, but if you grep for `engine` in the follow apps code files you'll see my engine code is similar to them. 

- x509.c
- srp.c
- verify.c
- pkey.c 
- rsautl.c
....and many others.

Comment (anhu) [reply apps/crl.c:207]: > I see this call without #ifdef in other apps...

I was just following what `OPENSSL_NO_MD5` ....if I follow that, it looks nicer and has less cognitive load.

**Comment by t8m** [comment]
It is fairly ugly to add an option to a command and immediately deprecate it. IMO this should be discussed in OTC.

**Comment by mattcaswell** [comment]
OTC: Engines can be loaded via the config file if needed. Engines are deprecated in 3.0 and therefore we will not be adding any further support.

**Comment by anhu** [comment]
Thanks for all the work in thinking about this.  

Note that 1.1.1 will get support as LTS  until 2023.  Presumably, that means that the next LTS branch will be cut in 2022; about a year from now.  I'm going to assume engines will not be removed by 2022 as that would be a massive undertaking because providers would have to take over everything, so that means the next LTS will likely have engine support.  If that branch is cut in 2022 and LTSs live for 5 years, you're letting this command-line inconsistency live until 2027. Does that really seem like a good idea to the OMC? 

The answer is likely yes and I'm fine with it.  Just wanted to make this timeline  a bit more explicit.

**Comment by mattcaswell** [comment]
The next LTS will be based on some version from 3.0 onwards. 3.0 deprecates all engines and they are not encouraged for use. Of course that does mean engines will still *exist* for some time to come - but we do expect that usage to be reducing. We are comfortable living with some inconsistencies until they are finally removed.

**Pull request closed without merging**